### PR TITLE
Stable diffusion inpainting.

### PR DESCRIPTION
**Pull request by patil-suraj**

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 1]
```diff
@@ -300,50 +289,57 @@ def __call__(
...
-        # preprocess image
-        if not isinstance(init_image, torch.FloatTensor):
-            init_image = preprocess_image(init_image)
+        # get the initial random noise unless the user supplied it
 
-        # encode the init image into latents and scale the latents
+        # Unlike in other pipelines, latents need to be generated in the target device
+        # for 1-to-1 results reproducibility with the CompVis implementation.
+        # However this currently doesn't work in `mps`.
+        latents_shape = (batch_size * num_images_per_prompt, 4, height // 8, width // 8)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:307]: The 4 is a bit too arbitrary here for me - we should save it somewhere in the model's config IMO

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 2]
```diff
@@ -354,17 +350,13 @@ def __call__(
...
-        # It's more optimized to move all timesteps to correct device beforehand
-        timesteps = self.scheduler.timesteps[t_start:].to(self.device)
-
-        for i, t in tqdm(enumerate(timesteps)):
+        for i, t in enumerate(self.progress_bar(timesteps_tensor)):
             # expand the latents if we are doing classifier free guidance
             latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
+
+            # concat latents, mask, masked_image in the channel dimension
+            latent_model_input = torch.cat([latent_model_input, mask, masked_image], dim=1)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:358]: agree that this shouldn't go into the unet_2d_condition model

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 3]
```diff
@@ -168,22 +164,13 @@ def __call__(
...
-                contain one color channel (L) instead of 3, so the expected shape would be `(B, H, W, 1)`.
-            strength (`float`, *optional*, defaults to 0.8):
-                Conceptually, indicates how much to inpaint the masked area. Must be between 0 and 1. When `strength`
-                is 1, the denoising process will be run on the masked area for the full number of iterations specified
-                in `num_inference_steps`. `init_image` will be used as a reference for the masked area, adding more
-                noise to that region the larger the `strength`. If `strength` is 0, no inpainting will occur.
+            height (`int`, *optional*, defaults to 512):
+                The height in pixels of the generated image.
+            width (`int`, *optional*, defaults to 512):
+                The width in pixels of the generated image.
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:170]:
```suggestion
                The width in pixels of the generated image.
            init_image (`torch.FloatTensor` or `PIL.Image.Image`):
                `Image`, or tensor representing an image batch, that will be used as the starting point for the
                process. This is the image whose masked region will be inpainted.
            mask_image (`torch.FloatTensor` or `PIL.Image.Image`):
                `Image`, or tensor representing an image batch, to mask `init_image`. White pixels in the mask will be
                replaced by noise and therefore repainted, while black pixels will be preserved. If `mask_image` is a
                PIL image, it will be converted to a single channel (luminance) before use. If it's a tensor, it should
                contain one color channel (L) instead of 3, so the expected shape would be `(B, H, W, 1)`.
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py (hunk 1 above)
Comment (anton-l) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:307]: something like `unet.in_channels - mask_channels - image_channels` should do the trick

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:368]: This input then has 9 channels?

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:368]: 4 + 2 for mask + 3 for masked_image ?

**Review by patrickvonplaten — APPROVED** [review]
Overall the PR looks very nice to me! 
Could we add one nice `ValueError` check / explanation in the beginning why the unet has 9 input channels? 
Guess it comes from 4 for the latent 3 for the masked image and 2 from the image or where does the 9 come from? Would be nice to wrap this info in a shape check / comment.

Apart from that good to go for me!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py (hunk 2 above)
Comment (anton-l) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:368]: Correct

File: tests/test_pipelines.py [hunk 4]
```diff
@@ -2059,7 +2217,7 @@ def test_stable_diffusion_img2img_onnx(self):
         assert np.abs(image_slice.flatten() - expected_slice).max() < 2e-2
 
     @slow
-    def test_stable_diffusion_inpaint_onnx(self):
+    def test_stable_diffusion_inpaint_legacy_onnx(self):
```
Comment (anton-l) [inline tests/test_pipelines.py:2220]:
```suggestion
    def test_stable_diffusion_inpaint_onnx(self):
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:325]: Done
